File: extract-download-images/rest/csharp/DownloadImages.cs
// Extracts all <img> src URLs from a page via the Browserless /scrape endpoint,
// then downloads each image to an images/ directory.
// Uses System.Net.Http.HttpClient and System.Text.Json — no packages needed.
//
// Run: dotnet run

using System;
using System.IO;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

class DownloadImages
{
    static async Task Main()
    {
        const string token = "YOUR_API_TOKEN_HERE";
        using var client = new HttpClient();

        var body = new StringContent(
            "{\"url\":\"https://example.com\",\"elements\":[{\"selector\":\"img\",\"timeout\":5000}]}",
            Encoding.UTF8,
            "application/json"
        );
        var scrapeRes = await client.PostAsync(
            $"https://production-sfo.browserless.io/scrape?token={token}",
            body
        );

        using var json = JsonDocument.Parse(await scrapeRes.Content.ReadAsStringAsync());
        var results = json.RootElement.GetProperty("data")[0].GetProperty("results");

        Directory.CreateDirectory("images");
        int i = 0;

        foreach (var result in results.EnumerateArray())
        {
            foreach (var attr in result.GetProperty("attributes").EnumerateArray())
            {
                string name = attr.GetProperty("name").GetString();
                string value = attr.GetProperty("value").GetString();

                if (name == "src" && value.StartsWith("http"))
                {
                    var bytes = await client.GetByteArrayAsync(value);
                    // Use the URL's file extension; fall back to .jpg for extensionless paths.
                    string ext = Path.GetExtension(new Uri(value).AbsolutePath) is { Length: > 0 } e ? e : ".jpg";
                    string filename = $"images/image-{i}{ext}";
                    await File.WriteAllBytesAsync(filename, bytes);
                    Console.WriteLine($"Saved {filename}");
                    i++;
                }
            }
        }
    }
}

File: examples/extract-download-images/rest/java/DownloadImages.java
// Extracts all <img> src URLs from a page via the Browserless /scrape endpoint,
// then downloads each image to an images/ directory.
// Uses java.net.http.HttpClient (Java 11+) and Gson for JSON parsing.
//
// Add to pom.xml:
//   <dependency>
//     <groupId>com.google.code.gson</groupId>
//     <artifactId>gson</artifactId>
//     <version>2.10.1</version>
//   </dependency>
//
// Compile: javac DownloadImages.java
// Run:     java DownloadImages

import com.google.gson.*;
import java.net.URI;
import java.net.http.*;
import java.nio.file.*;

public class DownloadImages {
    public static void main(String[] args) throws Exception {
        String token = "YOUR_API_TOKEN_HERE";
        HttpClient client = HttpClient.newHttpClient();

        String body = "{\"url\":\"https://example.com\",\"elements\":[{\"selector\":\"img\",\"timeout\":5000}]}";
        HttpRequest scrapeReq = HttpRequest.newBuilder()
            .uri(URI.create("https://production-sfo.browserless.io/scrape?token=" + token))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(body))
            .build();

        HttpResponse<String> scrapeRes = client.send(scrapeReq, HttpResponse.BodyHandlers.ofString());
        JsonObject json = JsonParser.parseString(scrapeRes.body()).getAsJsonObject();
        JsonArray results = json.getAsJsonArray("data").get(0)
            .getAsJsonObject().getAsJsonArray("results");

        Files.createDirectories(Path.of("images"));
        int i = 0;

        for (JsonElement resultEl : results) {
            for (JsonElement attrEl : resultEl.getAsJsonObject().getAsJsonArray("attributes")) {
                JsonObject attr = attrEl.getAsJsonObject();
                String name = attr.get("name").getAsString();
                String value = attr.get("value").getAsString();

                if ("src".equals(name) && value.startsWith("http")) {
                    HttpRequest imgReq = HttpRequest.newBuilder()
                        .uri(URI.create(value)).build();
                    HttpResponse<byte[]> imgRes = client.send(imgReq, HttpResponse.BodyHandlers.ofByteArray());

                    // Use URI.getPath() to strip query strings before extracting the extension.
                    String urlPath = URI.create(value).getPath();
                    int dot = urlPath.lastIndexOf('.');
                    String ext = dot >= 0 ? urlPath.substring(dot) : ".jpg";

                    Files.write(Path.of("images/image-" + i + ext), imgRes.body());
                    System.out.println("Saved images/image-" + i + ext);
                    i++;
                }
            }
        }
    }
}
